python-checks/src/main/java/org/sonar/python/checks/cdk/PublicNetworkAccessToCloudResourcesCheck.java

```diff
@@ -33,5 +62,109 @@ protected void registerFqnConsumer() {
+    Optional<CdkUtils.ExpressionFlow> associatedPublicIpAddress = getArgument(ctx, callExpression, ASSOCIATE_PUBLIC_IP_ADDRESS);
+    Optional<CdkUtils.ExpressionFlow> subnetId = getArgument(ctx, callExpression, "subnet_id");
+
+    if (associatedPublicIpAddress.filter(flow -> flow.hasExpression(isTrue())).isPresent()
+      && subnetId.filter(PublicNetworkAccessToCloudResourcesCheck::hasPrivateSubnetDefined).isEmpty()) {
+      associatedPublicIpAddress.get().addIssue(ERROR_MESSAGE);
+    }
+  }
+
+  // Check the *Sensitive network interface* as a specific DictionaryLiteral : '{"associate_public_ip_address" : True, "subnet_id" : *sensitive_or_no_subnet*}'
+  private static void checkNetworkInterfacesDictionary(SubscriptionContext ctx, Expression expression) {
+    List<CdkUtils.ResolvedKeyValuePair> elements = getDictionary(expression)
+      .map(dictionary -> resolveDictionary(ctx, dictionary))
+      .orElse(Collections.emptyList());
+
+    Optional<CdkUtils.ResolvedKeyValuePair> associatedPublicIpAddress = elements.stream().filter(el -> el.key.hasExpression(isString(ASSOCIATE_PUBLIC_IP_ADDRESS))).findFirst();
+    Optional<CdkUtils.ExpressionFlow> subnetId = elements.stream().filter(element -> element.key.hasExpression(isString("subnet_id"))).map(el -> el.value).findFirst();
+
+    if(associatedPublicIpAddress.filter(keyValuePair -> keyValuePair.value.hasExpression(isTrue())).isPresent()
+      && subnetId.filter(PublicNetworkAccessToCloudResourcesCheck::hasPrivateSubnetDefined).isEmpty()) {
+      associatedPublicIpAddress.get().key.getExpression(isString(ASSOCIATE_PUBLIC_IP_ADDRESS)).ifPresent(
+        expr -> raiseIssueOnParent(ctx, expr, Tree.Kind.KEY_VALUE_PAIR)
+      );
+    }
+  }
+
+  // Check the if the provided subnet is compliant : 'ec2.Vpc.select_subnets(subnet_type=ec2.SubnetType.PRIVATE_ISOLATED).subnet_ids[0]'
+  private static boolean hasPrivateSubnetDefined(CdkUtils.ExpressionFlow subnetId) {
+    return subnetId.getExpression(isSubscriptionExpression())
+      .map(SubscriptionExpression.class::cast).map(SubscriptionExpression::object)
+      .filter(expression -> expression.is(Tree.Kind.QUALIFIED_EXPR)).map(QualifiedExpression.class::cast)
+      .filter(PublicNetworkAccessToCloudResourcesCheck::isCompliantSubnet)
+      .isPresent();
+  }
+
+  private static boolean isCompliantSubnet(QualifiedExpression qualifiedExpression) {
+    return Optional.of(qualifiedExpression)
+      .filter(qualExpr -> qualExpr.name().name().equals("subnet_ids"))
+      .map(QualifiedExpression::qualifier)
+      .filter(expression -> expression.is(Tree.Kind.CALL_EXPR)).map(CallExpression.class::cast)
+      .filter(isFqn("aws_cdk.aws_ec2.Vpc.select_subnets"))
+      .flatMap(callExpression -> getArgument(null, callExpression, SUBNET_TYPE))
+      .filter(flow -> flow.hasExpression(isFqnOf(COMPLIANT_SUBNETS)))
+      .isPresent();
```

Comment:
does that mean we would only detect this issue when the subscription is done "inline"?

I guess we wouldn't detect an issue in the snippet below?

```python
sub = ec2.Vpc.select_subnets(subnet_type=ec2.SubnetType.PRIVATE_ISOLATED)
sub.subnet_ids[0] # FN?
```

The rule should be triggered only when `subnet_ids` is accessed? I guess an alternative could be to check if `subnet_type` argument has a `ec2.Vpc.select_subnets(subnet_type=ec2.SubnetType.PRIVATE_ISOLATED)` somewhere in the its children nodes?

I'm fine with it, just want to make sure it's intentional.
however let's document the FN with a unit test

Reply:
Indeed my approach was a bit too naive.
I changed and adapted the code so that it can handle different use case where the subnet_ids is accessed not directly :
`subnet_id = ec2.Vpc.select_subnets(subnet_type=ec2.SubnetType.PRIVATE_ISOLATED).subnet_ids[0]
        ec2.CfnInstance(network_interfaces=[ec2.CfnInstance.NetworkInterfaceProperty(associate_public_ip_address=True, subnet_id=subnet_id)])
        subnet_ids = ec2.Vpc.select_subnets(subnet_type=ec2.SubnetType.PRIVATE_ISOLATED).subnet_ids
        ec2.CfnInstance(network_interfaces=[ec2.CfnInstance.NetworkInterfaceProperty(associate_public_ip_address=True, subnet_id=subnet_ids[0])])
        subnet = ec2.Vpc.select_subnets(subnet_type=ec2.SubnetType.PRIVATE_ISOLATED)
        ec2.CfnInstance(network_interfaces=[ec2.CfnInstance.NetworkInterfaceProperty(associate_public_ip_address=True, subnet_id=subnet.subnet_ids[0])])
`